lib/services/local/plugin_data.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/gravitational/trace"
+)
+
+// PluginDataService manages PluginData
+type PluginDataService struct {
+	backend.Backend
+}
+
+// NewPluginDataService creates a new PluginDataService
+func NewPluginDataService(backend backend.Backend) *PluginDataService {
+	return &PluginDataService{backend}
+}
+
+// GetPluginData loads all plugin data matching the supplied filter.
+func (s *PluginDataService) GetPluginData(ctx context.Context, filter types.PluginDataFilter) ([]types.PluginData, error) {
+	data, err := s.getPluginData(ctx, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return data, nil
+}
+
+// getPluginData returns an array of PluginData matching filter criteria
+func (s *PluginDataService) getPluginData(ctx context.Context, filter types.PluginDataFilter) ([]types.PluginData, error) {
+	// Filters which specify Resource are a special case since they will match exactly zero or one
+	// possible PluginData instances.
+	if filter.Resource != "" {
+		item, err := s.Get(ctx, pluginDataKey(filter.Kind, filter.Resource))
+		if err != nil {
+			// A filter with zero matches is still a success, it just
+			// happens to return an empty slice.
+			if trace.IsNotFound(err) {
+				return nil, nil
+			}
+			return nil, trace.Wrap(err)
+		}
+		data, err := itemToPluginData(*item)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if !filter.Match(data) {
+			// A filter with zero matches is still a success, it just
+			// happens to return an empty slice.
+			return nil, nil
+		}
+		return []types.PluginData{data}, nil
+	}
+
+	prefix := backend.Key(pluginDataPrefix, filter.Kind)
+	result, err := s.GetRange(ctx, prefix, backend.RangeEnd(prefix), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var matches []types.PluginData
+	for _, item := range result.Items {
+		if !bytes.HasSuffix(item.Key, []byte(paramsPrefix)) {
+			// Item represents a different resource type in the
+			// same namespace.
+			continue
+		}
+		data, err := itemToPluginData(item)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if !filter.Match(data) {
+			continue
+		}
+		matches = append(matches, data)
+	}
+	return matches, nil
+}
+
+// UpsertsPluginData upserts a per-resource PluginData entry.
+func (s *PluginDataService) UpsertPluginData(ctx context.Context, params types.PluginDataUpdateParams, parentExpiresAt *time.Time) error {
+	return trace.Wrap(s.upsertPluginData(ctx, params, parentExpiresAt))
+}
+
+// updatePluginData updates or creates PluginData using CAS
+func (s *PluginDataService) upsertPluginData(ctx context.Context, params types.PluginDataUpdateParams, parentExpiresAt *time.Time) error {
+	retryPeriod := retryPeriodMs * time.Millisecond
+	retry, err := utils.NewLinear(utils.LinearConfig{
+		Step: retryPeriod / 7,
+		Max:  retryPeriod,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Update is attempted multiple times in the event of concurrent writes.
+	for i := 0; i < maxCmpAttempts; i++ {
```

Comment:
While this is being moved about, is it worth (or even a good idea...) changing this so that it keeps trying until the context expires, rather than some arbitrary number of retries?

Reply:
If there's some kind of exponential backoff then it's okay to rely on a context expiration. But it's better to handle `CompareFailed` on a client side and do the backoff there (like we do in `teleport-plugins`).

So whenever you don't have a backoff, you should definitely limit the try count cause otherwise you get the backend DoS'ed before the context expiration.

Actually, the `maxCmpAttempts = 1` would be enough but in order to reduce the network round trips of a client & its backoff, there're some multiple tries on a backend side too.